File: activity/20200904/RabbitLee/Minimum_Cost_For_Tickets.cpp
const int D = 365;
int DAYS[] = {1, 7, 30};
class Solution {
public:
    int mincostTickets(vector<int>& days, vector<int>& costs) {
        vector<int> f(D + 40, 0);
        unordered_set<int> s;
        for (int day : days) s.insert(day);
        for (int i = D; i >= 1; i--) {
            if (s.find(i) == s.end()) {
                f[i] = f[i + 1];
            } else {
                int ans = 1000 * D;
                for (int j = 0; j < 3; j++) {
                    ans = min(ans, costs[j] + f[i + DAYS[j]]);
                }
                f[i] = ans;
            }
        } 
        return f[1];
    }
};

File: activity/20200904/scau-flyingbird/MinimumCostForTickets.go
import "math"

func mincostTickets(days []int, costs []int) int {
	idx := []int{1, 7, 30}
	sum := make([]int, len(days))
	for i := 0; i < len(days); i++ {
		sum[i] = math.MaxInt32
	}
	return cal(0, &sum, &days, &costs, &idx)
}

func cal(i int, sum *[]int, days *[]int, costs *[]int, idx *[]int) int {
	if i >= len(*days) {
		return 0
	}

	if (*sum)[i] != math.MaxInt32 {
		return (*sum)[i]
	}

	min := math.MaxInt32
	for k := 0; k < len(*idx); k++ {
		j := i + 1
		for ; j < len(*days) && (*days)[j] < (*days)[i]+(*idx)[k]; j++ {

		}
		min = Min(min, cal(j, sum, days, costs, idx)+(*costs)[k])
	}
	(*sum)[i] = min
	return (*sum)[i]
}

func Min(a, b int) int {
	if a < b {
		return a
	} else {
		return b
	}
}
